### PR TITLE
Display image inside of indicator

### LINKED ISSUE #68: --grace 1 prevents screenlock
## Problem
The `--grace` option doesn't work as expected. The screen doesn't lock. To me it seems like the white default lock screen pops up for a split second but it never actually locks or instantly unlocks itself again without asking for a pw  

`swaylock -C /dev/null --grace 0`: screen locks as expected and I can unlock it by typing the password
`swaylock -C /dev/null --grace 1` (or any val > 0): screen seems to lock for a split second and instantly unlocks itself without asking for pw

Any idea why this is happening? Sth with pam maybe?

## Info
```bash
 ~> swaylock --version
swaylock version 1.7.0.0
 ~> hyprctl version
Hyprland, built from branch main at commit 9a09eac79b85c846e3a865a9078a3f8ff65a9259  (props: bump version to 0.42.0).
Date: 2024-08-07
Tag: v0.42.0, commits: 9a09eac79b85c846e3a865a9078a3f8ff65a9259

flags: (if any)

 ~> uname -a
Linux lt-t14 6.6.47 #1-NixOS SMP PREEMPT_DYNAMIC Mon Aug 19 04:04:32 UTC 2024 x86_64 GNU/Linux
```

## Trace
```Bash
 ~> swaylock -C /dev/null --trace --grace 0
2024-08-27 22:18:25 - [main.c:1870] Found config at /dev/null
2024-08-27 22:18:25 - [main.c:1880] Parsing CLI Args
2024-08-27 22:18:25 - [main.c:453]: trace: handle_wl_output_geometry
2024-08-27 22:18:25 - [main.c:657]: trace: handle_wl_output_name
2024-08-27 22:18:25 - [main.c:658] output name is eDP-1
2024-08-27 22:18:25 - [main.c:469]: trace: handle_wl_output_scale
2024-08-27 22:18:25 - [main.c:669]: trace: handle_wl_output_done
2024-08-27 22:18:25 - [main.c:453]: trace: handle_wl_output_geometry
2024-08-27 22:18:25 - [main.c:657]: trace: handle_wl_output_name
2024-08-27 22:18:25 - [main.c:658] output name is DP-4
2024-08-27 22:18:25 - [main.c:469]: trace: handle_wl_output_scale
2024-08-27 22:18:25 - [main.c:669]: trace: handle_wl_output_done
2024-08-27 22:18:25 - [main.c:453]: trace: handle_wl_output_geometry
2024-08-27 22:18:25 - [main.c:657]: trace: handle_wl_output_name
2024-08-27 22:18:25 - [main.c:658] output name is DP-5
2024-08-27 22:18:25 - [main.c:469]: trace: handle_wl_output_scale
2024-08-27 22:18:25 - [main.c:669]: trace: handle_wl_output_done
2024-08-27 22:18:25 - [main.c:535]: trace: handle_screencopy_frame_buffer
2024-08-27 22:18:25 - [main.c:535]: trace: handle_screencopy_frame_buffer
2024-08-27 22:18:25 - [main.c:535]: trace: handle_screencopy_frame_buffer
2024-08-27 22:18:25 - [main.c:562]: trace: handle_screencopy_frame_flags
2024-08-27 22:18:25 - [main.c:610]: trace: handle_screencopy_frame_ready
2024-08-27 22:18:25 - [main.c:562]: trace: handle_screencopy_frame_flags
2024-08-27 22:18:25 - [main.c:610]: trace: handle_screencopy_frame_ready
2024-08-27 22:18:25 - [main.c:562]: trace: handle_screencopy_frame_flags
2024-08-27 22:18:25 - [main.c:610]: trace: handle_screencopy_frame_ready
2024-08-27 22:18:25 - [main.c:1970] Using ext-session-lock-v1

 ~> swaylock -C /dev/null --trace --grace 1
2024-08-27 22:18:32 - [main.c:1870] Found config at /dev/null
2024-08-27 22:18:32 - [main.c:1880] Parsing CLI Args
2024-08-27 22:18:32 - [main.c:453]: trace: handle_wl_output_geometry
2024-08-27 22:18:32 - [main.c:657]: trace: handle_wl_output_name
2024-08-27 22:18:32 - [main.c:658] output name is eDP-1
2024-08-27 22:18:32 - [main.c:469]: trace: handle_wl_output_scale
2024-08-27 22:18:32 - [main.c:669]: trace: handle_wl_output_done
2024-08-27 22:18:32 - [main.c:453]: trace: handle_wl_output_geometry
2024-08-27 22:18:32 - [main.c:657]: trace: handle_wl_output_name
2024-08-27 22:18:32 - [main.c:658] output name is DP-4
2024-08-27 22:18:32 - [main.c:469]: trace: handle_wl_output_scale
2024-08-27 22:18:32 - [main.c:669]: trace: handle_wl_output_done
2024-08-27 22:18:32 - [main.c:453]: trace: handle_wl_output_geometry
2024-08-27 22:18:32 - [main.c:657]: trace: handle_wl_output_name
2024-08-27 22:18:32 - [main.c:658] output name is DP-5
2024-08-27 22:18:32 - [main.c:469]: trace: handle_wl_output_scale
2024-08-27 22:18:32 - [main.c:669]: trace: handle_wl_output_done
2024-08-27 22:18:32 - [main.c:535]: trace: handle_screencopy_frame_buffer
2024-08-27 22:18:32 - [main.c:535]: trace: handle_screencopy_frame_buffer
2024-08-27 22:18:32 - [main.c:535]: trace: handle_screencopy_frame_buffer
2024-08-27 22:18:32 - [main.c:562]: trace: handle_screencopy_frame_flags
2024-08-27 22:18:32 - [main.c:610]: trace: handle_screencopy_frame_ready
2024-08-27 22:18:32 - [main.c:562]: trace: handle_screencopy_frame_flags
2024-08-27 22:18:32 - [main.c:610]: trace: handle_screencopy_frame_ready
2024-08-27 22:18:32 - [main.c:562]: trace: handle_screencopy_frame_flags
2024-08-27 22:18:32 - [main.c:610]: trace: handle_screencopy_frame_ready
2024-08-27 22:18:32 - [main.c:1970] Using ext-session-lock-v1

### PR DESCRIPTION
**Pull request by ErikReider**
From https://github.com/mortie/swaylock-effects/pull/89

Fixes #68 
Requires #20 

Description:

Displays an image inside of the indicator behind the inside of the indicator (the clock, text, and its background)

Example:
`swaylock --indicator-image /PATH/TO/SOME/IMAGE`

or:

```dosini
indicator-image=/PATH/TO/SOME/IMAGE

fade-in=0.2
image=~/.cache/wallpaper
...
```
image

### CODE REVIEW

**Comment by jirutka** [comment]
Please resolve the file conflict.

**Comment by ErikReider** [comment]
> Please resolve the file conflict.

Yeah. Haven't really had that much time. Will during the weekend :)

**Comment by ErikReider** [comment]
Can't really test the changes due to #3

**Comment by ErikReider** [comment]
@jirutka It's ready now :)

File: render.c [hunk 1]
```diff
@@ -194,6 +200,34 @@ void render_frame(struct swaylock_surface *surface) {
 
 	if (state->args.indicator ||
 			(upstream_show_indicator && state->auth_state != AUTH_STATE_GRACE)) {
+		// Draw indicator image
+		cairo_surface_t * image = state->indicator_image;
+		if (image) {
+			int height = state->indicator_image_height;
+			int width = state->indicator_image_height;
```
Comment (jirutka) [inline render.c:207]: This is probably a typo.

File: render.c [hunk 2]
```diff
@@ -194,6 +200,34 @@ void render_frame(struct swaylock_surface *surface) {
...
+
+			// Create the arc that clips the image
+			cairo_arc(cairo,
+					buffer_diameter * 0.5,
+					buffer_diameter * 0.5,
+					radius,
+					0, 2 * M_PI);
+
+			// Scale cairo to make image fit the indicator
+			cairo_scale (cairo, scale, scale);
```
Comment (jirutka) [inline render.c:221]:
```suggestion
			cairo_scale(cairo, scale, scale);
```

File: render.c [hunk 3]
```diff
@@ -194,6 +200,34 @@ void render_frame(struct swaylock_surface *surface) {
...
+
+			cairo_set_source_surface(cairo, image, offset, offset);
+
+			// Scale cairo back
+			cairo_scale (cairo, 1 / scale, 1 / scale);
```
Comment (jirutka) [inline render.c:226]:
```suggestion
			cairo_scale(cairo, 1 / scale, 1 / scale);
```

File: render.c (hunk 1 above)
Comment (ErikReider) [reply render.c:207]: This might be because we want the image to be square? It was a while ago I coded this so might be a typo

**Comment by jirutka** [comment]
I’ve refactored it (#51) and merged as ea6688d8a56a271d99c9c801a32685ae98db1e5f.

**Pull request closed without merging**